### PR TITLE
Commit: show summary when commiting

### PR DESCRIPTION
**Pull request by yyyc514**
Relates to #107.

### CODE REVIEW

**Comment by yyyc514** [comment]
If this is heading the right direction I can improve the syntax further.

**Comment by yyyc514** [comment]
I bumped the help to `###` cause I thought it looked better.

File: core/commands/commit.py [hunk 1]
```diff
@@ -57,11 +59,37 @@ class GsCommitInitializeViewCommand(TextCommand, GitCommand):
...
+                filtered_lines.append(line)
+
+        return "\n".join(filtered_lines)
+
     def run(self, edit):
         merge_msg_path = os.path.join(self.repo_path, ".git", "MERGE_MSG")
 
-        option_amend = self.view.settings().get("git_savvy.commit_view.amend")
-        if option_amend:
+        if self.amending():
```
Comment (divmain) [inline core/commands/commit.py]: This is an unnecessary change.

File: core/commands/commit.py [hunk 2]
```diff
@@ -71,11 +99,14 @@ def run(self, edit):
...
 
         if sublime.load_settings("GitSavvy.sublime-settings").get("show_commit_diff"):
-            if option_amend:
+            if self.amending():
                 initial_text += self.git("diff", "HEAD^")
             else:
                 initial_text += self.git("diff", "--cached")
 
+        if sublime.load_settings("GitSavvy.sublime-settings").get("show_commit_summary"):
+            initial_text += self.fetch_commit_summary()
```
Comment (divmain) [inline core/commands/commit.py]: Just pass in `option_amend` here.

File: core/commands/commit.py [hunk 3]
```diff
@@ -57,11 +59,37 @@ class GsCommitInitializeViewCommand(TextCommand, GitCommand):
...
+        """
+
+        args = []
+        if self.amending():
+            args.append("--amend")
+
+        dryrun = self.git("commit", "--dry-run", *args, throw_on_stderr = False)
+        # detect trying to do a commit with nothing staged
+        if dryrun.find('no changes added') != -1:
+            sublime.error_message(NO_STAGED_FILES)
```
Comment (divmain) [inline core/commands/commit.py]: If you go this route, there should be an exit mechanism - after showing this error message, it'll continue one with executing this method.

Comment (yyyc514) [inline core/commands/commit.py]: I wasn't sure what the proper thing to do was since at that point we already have the view (it's created by the first command at the window level).

**Comment by divmain** [comment]
I think a more straightforward approach here would be something along the lines of the following.

```
status = self.git("status")
status_comment = "\n".join("## " + line for line in status.split("\n"))
```

Really, the normal behavior here for commits is to display the verbose `git status` output.

File: core/commands/commit.py (hunk 2 above)
Comment (yyyc514) [inline core/commands/commit.py]: Is there some reason we can't wrap it in a method vs passing it everywhere?  `amending` really does seem to be a property of `GsCommitInitializeViewCommand to me`.

File: core/commands/commit.py [hunk 4]
```diff
@@ -5,16 +5,18 @@
...
+### the commit, close the window. To sign off the commit press {key}-S.
 
-## You may also reference or close a GitHub issue with this commit.  To do so,
-## type `#` followed by the `tab` key.  You will be shown a list of issues
-## related to the current repo.  You may also type `owner/repo#` plus the `tab`
-## key to reference an issue in a different GitHub repo.
+### You may also reference or close a GitHub issue with this commit.  To do so,
+### type `#` followed by the `tab` key.  You will be shown a list of issues
+### related to the current repo.  You may also type `owner/repo#` plus the `tab`
+### key to reference an issue in a different GitHub repo.
```
Comment (divmain) [inline core/commands/commit.py]: I'd prefer to keep it at two `#`.

**Comment by yyyc514** [comment]
`status = self.git("status")`

There is no --amend option for status.  Since what we want is to know exactly what the commit is planning to do there is no substitute for asking the commit command itself.

**Comment by divmain** [comment]
Hmm that's a very good point @yyyc514 .  Let me think on this a bit longer.  Sorry if it feels like I took an axe to this - just got off a plane and pretty tired.  Very much appreciate the PR.  As much as possible I prefer to keep things a simple as possible until a clean abstraction becomes apparent, which is where most of my comments are coming from.  I'll re-review this tomorrow.

File: syntax/make_commit.YAML-tmLanguage [hunk 5]
```diff
@@ -6,7 +6,15 @@ uuid: 0d4c19fb-b197-4d90-806e-7212e542c86b
...
 - comment: commit header
   name: comment.git-savvy.make-commit
-  match: ^##.*
+  match: ^###.*
+- comment: blah
+  name: comment
+  begin: ^#[^#]
+  end: $
+  patterns:
+  - comment: git gile categories
```
Comment (divmain) [inline syntax/make_commit.YAML-tmLanguage]: `git file categories`?

Comment (yyyc514) [inline syntax/make_commit.YAML-tmLanguage]: Open to suggestions of a better name :)

Comment (divmain) [inline syntax/make_commit.YAML-tmLanguage]: I was just pointing out the typo :)

Comment (yyyc514) [inline syntax/make_commit.YAML-tmLanguage]: And I still didn't see it. LOL.  Ok.

File: core/commands/commit.py (hunk 2 above)
Comment (yyyc514) [inline core/commands/commit.py]: My next idea was to clean this up so it's just content = " ".[header(), help_text(), footer()].  Since lots of different pieces of the render need to know about amend it seems logical to make it accessible on the class vs passing it into every render method.

**Comment by divmain** [comment]
Alright, I'm going to hit the sack, but I'll take a look at this again tomorrow.  Thanks again.  One last nit-pick before I sign-off - the top method in each class should be `run`, for consistency.

**Comment by yyyc514** [comment]
> One last nit-pick before I sign-off - the top method in each class should be run, for consistency.

Oh sure, that's no problem.

**Comment by yyyc514** [comment]
I'll be more than happy to port this stuff to the new Interface framework when that is all polished up.

**Comment by yyyc514** [comment]
Rebased and squashed.

**Comment by divmain** [comment]
I'm traveling for the next few days and will want some time to dig into this before merging.  This PR will have to wait for a couple of days.

**Comment by yyyc514** [comment]
I've rebased and removed the error case.  Currently commit is not an error case if there are no files staged, so this probably shouldn't be either.  Plus if there are no files staged it's pretty easy to notice that when you're showing the dry-run.

Ping.

**Comment by divmain** [comment]
Not ignoring this one.  I'm mulling over how I want to approach the commit view, after working with the other interfaces.  Leaving this PR open and, even if it is not merged, I'll reference what you did here.

**Comment by stoivo** [comment]
Hey, how is the planing going

**Comment by divmain** [comment]
Closing this, but issue will be addressed as part of #50.
